src/D2L.CodeStyle.Analyzers/Language/DefaultValueConsistencyForOverriddenMethodsAnalyzer.cs

```diff
@@ -0,0 +1,135 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace D2L.CodeStyle.Analyzers.Language {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class DefaultValueConsistencyForOverriddenMethodsAnalyzer : DiagnosticAnalyzer {
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+			=> ImmutableArray.Create(
+				Diagnostics.IncludeDefaultValueInOverrideForReadability,
+				Diagnostics.DontIntroduceNewDefaultValuesInOverrides,
+				Diagnostics.DefaultValuesInOverridesShouldBeConsistent
+			);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+
+			context.RegisterSyntaxNodeAction(
+				AnalyzeMethodDeclaration,
+				SyntaxKind.MethodDeclaration
+			);
+		}
+
+		public static void AnalyzeMethodDeclaration(
+			SyntaxNodeAnalysisContext ctx
+		) {
+			var methodDecl = (MethodDeclarationSyntax)ctx.Node;
+
+			var hostDecl = methodDecl.Parent as TypeDeclarationSyntax;

```

Comment:
`.Parent` comes from the syntax base class. I'm assuming here that its a `TypeDeclarationSyntax` (particular subset of all syntax). I.e. every method decl has a parent of either a [class, struct or interface decl syntax](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Syntax/TypeDeclarationSyntax.cs,726b9d80aa76a768,references). I think that's true but haven't checked.

I should have used a stronger cast though. Will fix.

Reply:
Traced up through the parser and they will actually parse methods in some weird places just so they can continue on and find more diagnostics. For example:

![image](https://user-images.githubusercontent.com/597907/48277380-82d2c800-e418-11e8-989b-18d94b904f56.png)

So I'll keep the weak cast and bail if `null`.

Reply:
[Link to Roslyn code](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Parser/LanguageParser.cs,c55189f366e3bc94,references)